### PR TITLE
fix(components): [select] excute filter/remoteMethod when blur

### LINKED ISSUE #9539: [Component] [select] select自定义filter-method时的展现问题
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.37`
- Element Plus Version: `2.2.15`
- Browser / OS: `User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.5112.102 Safari/537.36 Edg/104.0.1293.63`
- Build Tool: `CDN`

### Related Component

- `el-select`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuXG5cbmNvbnN0IHNlbGVjdFZhbHVlID0gcmVmKClcbmNvbnN0IHNlbGVjdE9wdGlvbnM9cmVmKFtcbiAge1xuICAgIHZhbHVlOiAnT3B0aW9uMScsXG4gICAgbGFiZWw6ICdPcHRpb24xJyxcbiAgfSxcbiAge1xuICAgIHZhbHVlOiAnT3B0aW9uMicsXG4gICAgbGFiZWw6ICdPcHRpb24yJyxcbiAgfSxcbiAge1xuICAgIHZhbHVlOiAnT3B0aW9uMycsXG4gICAgbGFiZWw6ICdPcHRpb24zJyxcbiAgfSxcbiAge1xuICAgIHZhbHVlOiAnT3B0aW9uNCcsXG4gICAgbGFiZWw6ICdPcHRpb240JyxcbiAgfSxcbiAge1xuICAgIHZhbHVlOiAnT3B0aW9uNScsXG4gICAgbGFiZWw6ICdPcHRpb241JyxcbiAgfSxcbl0pXG5jb25zdCBzZWxlY3RPcHRpb25zQ29weSA9IEpTT04ucGFyc2UoSlNPTi5zdHJpbmdpZnkoc2VsZWN0T3B0aW9ucy52YWx1ZSkpXG5cbmNvbnN0IHRlc3RmaWx0ZXIgPSAodGV4dCkgPT4ge1xuICBjb25zb2xlLmxvZygnc2VhcmNoOiAnICsgdGV4dClcblx0aWYodGV4dCkge1xuICAgIHNlbGVjdE9wdGlvbnMudmFsdWUgPSBzZWxlY3RPcHRpb25zQ29weS5maWx0ZXIoKGl0ZW0pID0+IHtcbiAgICAgIHJldHVybiBpdGVtLmxhYmVsLmluZGV4T2YodGV4dCkgPiAwXG4gICAgfSlcblx0fSBlbHNlIHtcblx0XHRzZWxlY3RPcHRpb25zLnZhbHVlID0gc2VsZWN0T3B0aW9uc0NvcHlcblx0fVxufVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgICA8ZGl2PlxuICAgICAgPGI+RWxTZWxlY3Toh6rlrprkuYlmaWx0ZXItbWV0aG9k6Zeu6aKY77yaPC9iPjxicj48YnI+XG4gICAgKDEpIOeCueWHu+W5tuW9leWFpeS7u+aEj+aQnOe0ouWtl+espuS4sjxicj5cbiAgICAoMikg56e75Ye654Sm54K5PGJyPlxuICAgICgzKSDlho3mrKHngrnlh7vvvIzkuIvmi4npgInpobnku43kuLrkuIrmrKHnmoTmkJzntKLnu5PmnpzvvIjlupTmmL7npLrmiYDmnInpgInpobnvvIzmiJbkv53nlZnkuIrmrKHnmoTmkJzntKLlrZfnrKbkuLLvvIk8YnI+PGJyPlxuICA8L2Rpdj5cbiAgPGVsLXNlbGVjdCByZWY9XCJ0ZXN0XCIgdi1tb2RlbD1cInNlbGVjdFZhbHVlXCIgZmlsdGVyYWJsZSBwbGFjZWhvbGRlcj1cIlNlbGVjdFwiIDpmaWx0ZXItbWV0aG9kPVwidGVzdGZpbHRlclwiIDptdWx0aXBsZT1cInRydWVcIj5cblx0XHRcdDxlbC1vcHRpb25cblx0XHRcdCAgdi1mb3I9XCJpdGVtIGluIHNlbGVjdE9wdGlvbnNcIlxuXHRcdFx0ICA6a2V5PVwiaXRlbS52YWx1ZVwiXG5cdFx0XHQgIDpsYWJlbD1cIml0ZW0ubGFiZWxcIlxuXHRcdFx0ICA6dmFsdWU9XCJpdGVtLnZhbHVlXCJcblx0XHRcdC8+XG5cdFx0ICA8L2VsLXNlbGVjdD5cblxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydF9tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHt9XG59IiwiX28iOnt9fQ==)

### Steps to reproduce

1 为select自定义filter-method（以下简称filter）
2 在连续编辑搜索字符串时，filter能够一直被正确调用
3 从select移出焦点再移入

### What is Expected?

下拉选项仍是之前的过滤结果，保留之前的搜索字符串；或展示全部选项并清空搜索字符串

### What is actually happening?

下拉选项仍是之前的过滤结果，同时也未保留之前的搜索字符串，二者矛盾

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by yuanxin518**
closed #9539

when select component trigger blur event, it need excute the filterMethod or remoteMethod,
otherwise, the next time you click select, you may filter the result to the previous one.

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by Circkoooooo** [comment]
#9539

**Comment by Circkoooooo** [comment]
@xiaoxian521 hi, is there anything needs to be modified? Hope you help to approving some workflow as it's  my first contribution  for ele.

**Comment by Circkoooooo** [comment]
@sxzz Can you help review it?

File: packages/components/select/src/useSelect.ts [hunk 1]
```diff
@@ -225,6 +225,14 @@ export const useSelect = (props, states: States, ctx) => {
     () => states.visible,
     (val) => {
       if (!val) {
+        if (props.filterable) {
+          if (typeof props.filterMethod === 'function') {
```
Comment (tolking) [inline packages/components/select/src/useSelect.ts:229]:
```suggestion
          if (isFunction(props.filterMethod)) {
```

Maybe use the `isFunction` functions from @element-plus/utils

Comment (tolking) [inline packages/components/select/src/useSelect.ts:229]:
```suggestion
          if (isFunction(props.filterMethod)) {
```

Maybe use the `isFunction` functions from @element-plus/utils

Comment (Circkoooooo) [reply packages/components/select/src/useSelect.ts:229]: `typeof ...'function'` also used in else where in this file, so had better optimise it all by others?

Comment (sxzz) [reply packages/components/select/src/useSelect.ts:229]: I think you only need to modify your code that fixes the bug you describe. The others can be refactored in a new PR.

File: packages/components/select/src/useSelect.ts [hunk 2]
```diff
@@ -225,6 +225,14 @@ export const useSelect = (props, states: States, ctx) => {
...
+            props.filterMethod()
+          }
+          if (typeof props.remoteMethod === 'function') {
```
Comment (sxzz) [inline packages/components/select/src/useSelect.ts:232]: Same here.

File: packages/components/select/src/useSelect.ts (hunk 1 above)
Comment (Circkoooooo) [reply packages/components/select/src/useSelect.ts:229]: > I think you only need to modify your code that fixes the bug you describe. The others can be refactored in a new PR.

yes i fixed it, so can i close it?

Comment (sxzz) [reply packages/components/select/src/useSelect.ts:229]: I didn't see it. Did you push the code to github?

Comment (Circkoooooo) [reply packages/components/select/src/useSelect.ts:229]: > I didn't see it. Did you push the code to github?
Have  you seen the commit code `36c2286`, it's my commit.

Comment (Circkoooooo) [reply packages/components/select/src/useSelect.ts:229]: > I didn't see it. Did you push the code to github?

图片
the mark one is my code and the next one is a extra commit.

Comment (tolking) [reply packages/components/select/src/useSelect.ts:229]: > `typeof ...'function'` also used in else where in this file, so had better optimise it all by others?

Other files do not need to be considered. but for your changes, you should write in the format we recommend.

Comment (Circkoooooo) [reply packages/components/select/src/useSelect.ts]: > > `typeof ...'function'` also used in else where in this file, so had better optimise it all by others?
> 
> Other files do not need to be considered. but for your changes, you should write in the format we recommend.

I've already made changes in the latest commit, please check to see.

**Review by tolking — APPROVED** [review]

**Comment by Circkoooooo** [comment]
@tolking Need approve.🙄

**Pull request merged**